### ISSUE TITLE
Support for hostname canonicalization in ssh client config

### ISSUE THREAD

**Issue by siliconblade** [ronf/asyncssh#713]
Currently asyncssh does not support hostname canonicalization in the client config [[docs](https://asyncssh.readthedocs.io/en/stable/api.html#config-file-support)].  Please add support for the following client options that are present in openssh:

`CanonicalizeHostname yes
CanonicalDomains example.org example.com
CanonicalizeMaxDots 1
CanonicalizeFallbackLocal yes
CanonicalizePermittedCNAMEs *.redis.example.org:redis.example.com`

Paramiko has [implemented](https://github.com/paramiko/paramiko/blob/main/paramiko/config.py) these partially.

**Comment by ronf** [comment]
I'll look into this. It appears to be pretty straightforward, but there's a lot of new config as you pointed out here, plus an additional "Match canonical" option in the config file and a need to reload the config whenever canonicalization happens.

In the meantime, you can get the same result as this by just creating "Host" entries in your config file with both the short and long name. For instance:

```
Host host1 host1.example.com
    Hostname host1.example.com

Host host2 host2.example.com
    Hostname host2.example.com

Host host3 host3.example.net
    Hostname host3.example.net
```

With this config, you can use either the short or long name as the target host when running SSH. The hostname it looks up in DNS will be whatever hostname you put in the Hostname option. You can even use IP addresses there if the target host doesn't have an entry in DNS but you still want to refer to it using a name. In that sense, it's more flexible than hostname canonicalization, but it has to be set up on a per-host basis.

**Comment by siliconblade** [comment]
Thank you @ronf for looking into this. Unfortunately there are a ton of hosts that need full names and there can be multiple domains to consider. We have a workaround for this, but having configs in multiple places adds to user confusion.

**Comment by ronf** [comment]
Yeah, understood.

I made some good progress today on this. I have all the config and options pieces done to pass in the new configuration, and a first cut at the canonicalization function itself. I'm still debugging that and additional work will be required to handle the case where SSH tunnels are involved, and to handle the enforcement of CanonicalizePermittedCNAMEs to decide whether to return a CNAME or not, but it's a start!

I've also got documentation and unit tests to write once I get the basic functionality working, so it may be a week or two before this is fully done. However, if you're interested in trying out a preview of it, I could potentially make a first cut available once I have something working.

**Comment by siliconblade** [comment]
Definitely a great start and more than happy to test early versions. Thank you!

**Comment by ronf** [comment]
Ok - a first cut of this is now available as commit b99e8d5 in the "develop" branch. It supports all five of the config parameters from OpenSSH, and also adds support for "canonical" and "final" match keywords. Thanks to @commonism, negation of match keywords is also now supported.

I also added "host" and "port" as new items in `get_extra_info()` on `SSHConnection`, so you can query what the hostname ends up being after canonicalization.

This feature could probably still use a few more unit tests, but what's here is enough to at least provide full coverage of all the new code. Let me know if you run into any issues with it.

**Comment by siliconblade** [comment]
Thanks @ronf , I tested the dev branch and had to update [connection.py](https://github.com/ronf/asyncssh/blob/edc96733c08e9d2a3dbaf432b6192f1c9217d69a/asyncssh/connection.py#L277)  to look like the following for it to work, which is more similar to what [openssh has](https://github.com/openssh/openssh-portable/blob/8eabd2ae2ca1d7756417a1ee5b41f09c5d997634/ssh.c#L253). The issue I had arises when my DNS resolver returns a bogus IP address when it can't resolve the hostname. 

One other thing that did not work was having 2 separate "Host *" entries in the config with different settings (i.e. one with Canonicalization settings, the other with username, key path, port). I had to merge the separate "Host *" entries for the configs in the second entry to get recognized.

```
async def _resolve_host(host, port, loop: asyncio.AbstractEventLoop) -> Optional[str]:

   """Attempt to resolve a hostname, returning a canonical name"""

   try:
       addrinfo = await loop.getaddrinfo(
           host + ".", port, family=socket.AF_UNSPEC, type=socket.SOCK_STREAM, proto=0
       )
   except socket.gaierror:
       return None
   else:
       return addrinfo[0][3]


async def _canonicalize_host(loop: asyncio.AbstractEventLoop,
                            options: 'SSHConnectionOptions') -> Optional[str]:
   """Canonicalize a host name"""

   host = options.host
   port = options.port

   if not options.canonicalize_hostname or not options.canonical_domains or \
           host.count('.') > options.canonicalize_max_dots or \
           (await _resolve_host(host, port, loop)):
       return None

   for domain in options.canonical_domains:
       canon_host = f'{host}.{domain}'
       cname = await _resolve_host(canon_host, port, loop)

       if cname is not None:
           if cname:
               for patterns in options.canonicalize_permitted_cnames:
                   host_pat, cname_pat = map(WildcardPatternList, patterns)

                   if host_pat.matches(canon_host) and \
                           cname_pat.matches(cname):
                       canon_host = cname
                       break

           return canon_host

   if not options.canonicalize_fallback_local:
       raise OSError(f'Unable to canonicalize hostname "{host}"')

   return None
```

**Comment by ronf** [comment]
Thanks for testing this. A few questions:

In _resolve_host, it looks like you added "proto=0", but according to the documentation that should already be the default for that argument. Similarly, family defaults to 0, which matches with socket.AF_UNSPEC, so it seems like leaving those two arguments out shouldn't change anything. Did they make a difference for you, or does it work if you just include `type=socket.SOCK_STREAM`? I have seen `getaddrinfo()` return different results based on including type or not, but I haven't specifically seen that affect the CNAME returned, which is the only thing that really matters here.

The other change in _resolve_host appears to be passing in the original port rather than 0. Does that actually change the set of hostnames which are returned? It doesn't seem to for me on macOS.

I don't see any other changes -- did I catch everything?

I've tried to reproduce the issue you mentioned with multiple "Host *" blocks, but so far I haven't seen the problem. When I have two such blocks, it seems to work with the blocks in either order. Can you give me a more specific example where you're seeing this problem, including both what's in the config and what's in the `connect()` call?

**Comment by siliconblade** [comment]
Thanks for the quick response. 

The documentation mentions that the defaults can be system specific so I provided values explicitly similar to openssh. The actual difference in results it seems happened due to removing flags=socket.AI_CANONNAME.

I can't seem to replicate the config issue unfortunately, now I'm running into "Host key is not trusted for host..." and it looks like "StrictHostKeyChecking no" config is not available.

**Comment by ronf** [comment]
You're right that StrictHostKeyChecking is not supported. See #685 for some discussion on this. It basically comes down to AsyncSSH never modifying OpenSSH configuration files, and AsyncSSH never prompting the user, both of which would be required to support some of the StrictHostKeyChecking modes.

You can set `known_hosts=None` (if you are willing to deal with the man-in-the-middle risk), or you can do this through the `~/.ssh/config` by setting UserKnownHostsFile to "none".

Regarding the flags, setting socket.AI_CANONNAME flag is needed to get the resolver to return CNAME values, Without this flag set, the CNAME functionality of OpenSSH can't work. In some cases, this may result in a CNAME being found but no corresponding address for that name. Could that be what you're running into?

**Comment by siliconblade** [comment]
Thank you for following up, hopefully the debug output from openssh and from asynssh helps:

```
% ssh -vvv server01
OpenSSH_9.8p1, LibreSSL 3.3.6
debug1: Reading configuration data /Users/myuser/.ssh/config
debug1: /Users/myuser/.ssh/config line 1: Applying options for *
debug1: /Users/myuser/.ssh/config line 7: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 21: include /etc/ssh/ssh_config.d/* matched no files
debug1: /etc/ssh/ssh_config line 54: Applying options for *
debug2: resolve_addr: could not resolve name server01 as address: nodename nor servname provided, or not known
debug3: resolve_canonicalize: attempting "server01" => "[server01.x.example.com](http://server01.x.example.com/)."
debug3: resolve_host: lookup server01.x.example.com.:2345
debug1: Canonicalized hostname "server01" => "[server01.x.example.com](http://server01.x.example.com/)"
```

```
% python tests.py
2024-11-28 13:40:11 DEBUG    selector_events - Using selector: KqueueSelector
2024-11-28 13:40:12 DEBUG    logging - Reading config from "/Users/myuser/.ssh/config"
2024-11-28 13:40:12 DEBUG    logging - _resolve_host: host server01 resolved to cname server01 and  address ('143.244.220.150', 0)
2024-11-28 13:40:12 INFO     logging - Opening SSH connection to server01, port 2345
```

openssh first tries with resolve_addr (flags=AI_NUMERICHOST|AI_NUMERICSERV), which fails for server01, and then tries resolve_host (flags=AI_CANONNAME) with canonical name server01.x.example.com. 

My DNS setup is probably not the healthiest, but basically asyncssh's resolve_host will be presented with ATT's DNS error helper (143.244.220.150) when the lookup fails instead of erroring out.

**Comment by ronf** [comment]
Since AsyncSSH only supports numeric port numbers, I don't think socket.AI_NUMERICSERV will be needed here, but I can see why OpenSSH might support it to allow string service names that get mapped to a port number.

I think the real key here is socket.AI_NUMERICHOST. I didn't realize that OpenSSH didn't try to look up the hostname before canonicalizing it. I saw it doing a resolution in the debug output, and assumed it was trying to see if it could match the hostname as a fully-qualified name before trying to add any of the canonical domains to it. If it just wants to support literal (numeric) IP addresses in that step, that would actually make things simpler in some ways. I might even be able to avoid the expense of going out to the resolver at all for that.

I'll try and put together a diff you can test which skips the initial lookup and see if that helps.

I also need to go back and some extra logging around this feature (mostly at the debug level, but I think an info level message when it canonicalizes a host would also be worthwhile. I'll do that as a separate change after this issue is worked out.

**Comment by ronf** [comment]
Here's an updated version to try (meant to apply to the latest "develop" branch code.

```patch
diff --git a/asyncssh/connection.py b/asyncssh/connection.py
index e59fff6..3c666c0 100644
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -25,6 +25,7 @@ import functools
 import getpass
 import inspect
 import io
+import ipaddress
 import os
 import shlex
 import socket
@@ -274,18 +275,6 @@ _DEFAULT_LINE_HISTORY = 1000        # 1000 lines
 _DEFAULT_MAX_LINE_LENGTH = 1024     # 1024 characters


-async def _resolve_host(host, loop: asyncio.AbstractEventLoop) -> Optional[str]:
-    """Attempt to resolve a hostname, returning a canonical name"""
-
-    try:
-        addrinfo = await loop.getaddrinfo(host + '.', 0,
-                                          flags=socket.AI_CANONNAME)
-    except socket.gaierror:
-        return None
-    else:
-        return addrinfo[0][3]
-
-
 async def _canonicalize_host(loop: asyncio.AbstractEventLoop,
                              options: 'SSHConnectionOptions') -> Optional[str]:
     """Canonicalize a host name"""
@@ -293,25 +282,35 @@ async def _canonicalize_host(loop: asyncio.AbstractEventLoop,
     host = options.host

     if not options.canonicalize_hostname or not options.canonical_domains or \
-            host.count('.') > options.canonicalize_max_dots or \
-            (await _resolve_host(host, loop)):
+            host.count('.') > options.canonicalize_max_dots:
+        return None
+
+    try:
+        ipaddress.ip_address(host)
+    except ValueError:
+        pass
+    else:
         return None

     for domain in options.canonical_domains:
         canon_host = f'{host}.{domain}'
-        cname = await _resolve_host(canon_host, loop)

-        if cname is not None:
-            if cname:
-                for patterns in options.canonicalize_permitted_cnames:
-                    host_pat, cname_pat = map(WildcardPatternList, patterns)
+        try:
+            addrinfo = await loop.getaddrinfo(
+                canon_host, 0, flags=socket.AI_CANONNAME)
+        except socket.gaierror:
+            continue
+
+        cname = addrinfo[0][3]
+
+        if cname:
+            for patterns in options.canonicalize_permitted_cnames:
+                host_pat, cname_pat = map(WildcardPatternList, patterns)

-                    if host_pat.matches(canon_host) and \
-                            cname_pat.matches(cname):
-                        canon_host = cname
-                        break
+                if host_pat.matches(canon_host) and cname_pat.matches(cname):
+                    return cname

-            return canon_host
+        return canon_host

     if not options.canonicalize_fallback_local:
         raise OSError(f'Unable to canonicalize hostname "{host}"')
```

**Comment by ronf** [comment]
Unfortunately, if your DNS server always returns an address even when the hostname isn't found, I think that will end up breaking the ability to try multiple canonical domains. It'll end up always matching the first one in your list and not try the others, and I don't think there's anything I can do in AsyncSSH to improve on that.

This will also mean that hostnames which shouldn't get canonicalized probably will, even when the hostname lookup should have failed for all of the domains, and it'll never fall back to the legacy resolver (which may have different rules about what domains to search).

**Comment by ronf** [comment]
Great! This is now available as commit 9a3d13f in the "develop" branch. Thanks for your help in validating this!

**Comment by ronf** [comment]
Logging support for hostname canonicalization is now available in commit 6ecb91e.

**Comment by siliconblade** [comment]
Nice! Thank you for all the great work.

**Comment by ronf** [comment]
My pleasure -- this feature is now available in AsyncSSH 2.19.0.
